qa/rolling-upgrade/src/javaRestTest/java/org/elasticsearch/upgrades/FileSettingsRoleMappingUpgradeIT.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.upgrades;
+
+import com.carrotsearch.randomizedtesting.annotations.Name;
+
+import org.elasticsearch.client.Request;
+import org.elasticsearch.core.SuppressForbidden;
+import org.elasticsearch.test.XContentTestUtils;
+import org.elasticsearch.test.cluster.ElasticsearchCluster;
+import org.elasticsearch.test.cluster.local.distribution.DistributionType;
+import org.elasticsearch.test.cluster.util.Version;
+import org.elasticsearch.test.cluster.util.resource.Resource;
+import org.elasticsearch.test.junit.RunnableTestRuleAdapter;
+import org.junit.ClassRule;
+import org.junit.rules.RuleChain;
+import org.junit.rules.TemporaryFolder;
+import org.junit.rules.TestRule;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.function.Supplier;
+
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.not;
+import static org.hamcrest.Matchers.nullValue;
+
+public class FileSettingsRoleMappingUpgradeIT extends ParameterizedRollingUpgradeTestCase {
+
+    private static final RunnableTestRuleAdapter versionLimit = new RunnableTestRuleAdapter(
+        () -> assumeTrue(
```

Comment:
This is kind of related to the comment [here](https://github.com/elastic/elasticsearch/blob/8.x/x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/AbstractUpgradeTestCase.java#L52-L60), that was removed in 9.0.

```
    @Deprecated(forRemoval = true)
    @UpdateForV9
    // Tests should be reworked to rely on features from the current cluster (old, mixed or upgraded).
    // Version test against the original cluster will be removed
    protected static boolean isOriginalClusterVersionAtLeast(Version supportedVersion) {
        // Always assume non-semantic versions are OK: this method will be removed in V9, we are testing the pre-upgrade cluster version,
        // and non-semantic versions are always V8+
        return parseLegacyVersion(UPGRADE_FROM_VERSION).map(x -> x.onOrAfter(supportedVersion)).orElse(true);
    }

```

Reply:
Yeah, thats a good idea. You could also use synthetic version features to reference specific release versions

Reply:
I'm up for any approach here -- I copied the version check from the existing `FileSettingsUpgradeIT` test suite. 

What I need to check for is that we have a cluster that understands file settings (`8.4.0` and up) but is before the version in which we introduced the buggy behavior (prior to `8.15.0`, or more precisely prior to [this](https://github.com/elastic/elasticsearch/blob/37f03dc40dfcc55cc08fdbb6d30079f74b230710/server/src/main/java/org/elasticsearch/TransportVersions.java#L116) transport version). 

@thecoop do you have a pointer for how I could best capture this with synthetic version features/a test only feature?

Reply:
`oldClusterHasFeature("gte_v8.4.0") && oldClusterHasFeature("gte_v8.15.0") == false`

Reply:
oh wow that's neat. I'll do that 👍

Reply:
One note on this is that `oldClusterHasFeature` does not seem to work in `RunnableTestRuleAdapter` because there is a static variable that only initialized in a non-static context: https://github.com/elastic/elasticsearch/blob/075cd24043bcd718c45ba61963d364df7c6c7862/x-pack/plugin/esql/qa/server/mixed-cluster/src/javaRestTest/java/org/elasticsearch/xpack/esql/qa/mixed/MixedClusterEsqlSpecIT.java#L46

Reply:
Eh, yes, in that case you can define a historical feature for 8.4.0, and a test feature for 8.15.0, and check between the two

Reply:
I reworked it to not use a `RunnableTestRuleAdapter` and instead have an `assumeTrue` in a `Before` -- seems sufficient but let me know if I should go for the alternative you're suggesting instead